cirq/sim/sparse_simulator.py

```diff
@@ -201,59 +202,93 @@ def _run(
                 seed=self._prng,
             )
 
-        qid_shape = protocols.qid_shape(qubit_order)
-        intermediate_state = step_result.state_vector().reshape(qid_shape)
         return self._brute_force_samples(
-            initial_state=intermediate_state,
+            acton_args=acton_args,
             circuit=general_suffix,
             repetitions=repetitions,
-            qubit_order=qubit_order,
+            qubits=qubits,
         )
 
     def _brute_force_samples(
         self,
-        initial_state: np.ndarray,
+        acton_args: act_on_state_vector_args.ActOnStateVectorArgs,
         circuit: circuits.Circuit,
-        qubit_order: 'cirq.QubitOrderOrList',
+        qubits: Tuple['cirq.Qid', ...],
         repetitions: int,
     ) -> Dict[str, np.ndarray]:
         """Repeatedly simulate a circuit in order to produce samples."""
 
         measurements: DefaultDict[str, List[np.ndarray]] = collections.defaultdict(list)
         for _ in range(repetitions):
-            all_step_results = self._base_iterator(
-                circuit, initial_state=initial_state, qubit_order=qubit_order
+            all_step_results = self._core_iterator(
+                circuit, sim_state=acton_args.copy(), qubits=qubits
```

Comment:
I can't find the specific place, but somewhere inside the old code's `_base_iterator`, `qis._to_valid_state_vector` is making a copy. This is how _brute_force_samples is designed to work: it takes the intermediate state produced by the unitary prefix, and runs repetitions on copies of it. If state was not copied for each repetition, we'd see the repetitions affect each other.

The new code calls `_core_iterator` instead of `_base_iterator`, avoiding that call to `qis._to_valid_state_vector`, so avoids that extra copy.